crates/bevy_ui/src/update.rs

```diff
@@ -64,6 +65,30 @@ fn update_hierarchy(
     current_global_z
 }
 
+/// Correct the `ComputedVisibility` set by [`check_visibility`] for UI nodes.
+///
+/// Since [`check_visibility`] has no concept of UI cameras, we need a "post-pass"
+/// where we re-enable UI nodes that are visible because they have a matching
+/// UI camera in their layer.
+///
+/// [`check_visibility`]: bevy_render::view::visibility::check_visibility
+pub fn update_layer_visibility(
+    ui_cams: Query<&UiCameraConfig>,
+    mut nodes: Query<(&mut ComputedVisibility, &RenderLayers), With<Node>>,
+) {
+    for config in &ui_cams {
+        // Skip configs with default render layers, since `check_visibility` assumes it
+        if config.ui_render_layers == RenderLayers::default() {
+            continue;
+        }
+        for (mut visibility, node_layers) in &mut nodes {
+            if config.ui_render_layers.intersects(node_layers) {
+                visibility.set_visible_in_view();
```

Comment:
Will this override all other forms of computed visibility?

Reply:
I think it does, which will be an issue with inherited visibility. Good catch, need to update the code.

I don't like this, though. The default visibility check is buggy on UI entities since it cannot account for `UiCameraConfig`, but it does the inherited visibility thing.

In practice, that means the checks in `check_visibility` needs to be completely discarded and recomputed here :/

Reply:
Yeah, why not merge this together with those checks in some way?

Reply:
This would create a circular dependency between `bevy_render` and `bevy_ui` (need access to the `UiCameraConfig` and `Node` components, and you wouldn't want `bevy_render` to depend on `bevy_ui` regardless).

An alternative design where the UI camera is spawned as any other camera in the app world is possible, but the fact that it's a sort of special camera that shares a viewport with another one makes it awkward. I've given motivation in the opening comment of #5252.

Hot take: should just remove `Visibility` and `ComputedVisibility` from UI entities. Bonus: avoids confusion with `display: None` :D

But I feel like this is a discussion that should be had in a different PR or an issue.

Reply:
Oof, right :(